File: web/src/shamsi/solarHijri.ts
// Solar Hijri (هجری شمسی) <-> Gregorian conversion — the TypeScript port of the
// Android app's core:common/time/SolarHijri.kt (jalaali break-year algorithm).
// The business calendar of WorkTrack is Solar Hijri; storage/API stay ISO.

export interface ShamsiDate {
  year: number;
  month: number; // 1..12, 1 = Hamal/حمل
  day: number;
}

const BREAKS = [
  -61, 9, 38, 199, 426, 686, 756, 818, 1111, 1181, 1210, 1635, 2060, 2097, 2192,
  2262, 2324, 2394, 2456, 3178,
];

interface JalCal {
  leap: number;
  gy: number;
  march: number;
}

function jalCal(jy: number): JalCal {
  const gy = jy + 621;
  let leapJ = -14;
  let jp = BREAKS[0];

  let jump = 0;
  for (let i = 1; i < BREAKS.length; i++) {
    const jm = BREAKS[i];
    jump = jm - jp;
    if (jy < jm) break;
    leapJ += Math.floor(jump / 33) * 8 + Math.floor((jump % 33) / 4);
    jp = jm;
  }
  let n = jy - jp;

  leapJ += Math.floor(n / 33) * 8 + Math.floor(((n % 33) + 3) / 4);
  if (jump % 33 === 4 && jump - n === 4) leapJ += 1;

  const leapG = Math.floor(gy / 4) - Math.floor((Math.floor(gy / 100) + 1) * 3 / 4) - 150;
  const march = 20 + leapJ - leapG;

  if (jump - n < 6) n = n - jump + Math.floor((jump + 4) / 33) * 33;
  let leap = (((n + 1) % 33) - 1) % 4;
  if (leap === -1) leap = 4;

  return { leap, gy, march };
}

function g2d(gy: number, gm: number, gd: number): number {
  let d =
    Math.floor((gy + Math.floor((gm - 8) / 6) + 100100) * 1461 / 4) +
    Math.floor((153 * ((gm + 9) % 12) + 2) / 5) +
    gd -
    34840408;
  d = d - Math.floor((Math.floor((gy + 100100 + Math.floor((gm - 8) / 6)) / 100) * 3) / 4) + 752;
  return d;
}

function d2g(jdn: number): { gy: number; gm: number; gd: number } {
  let j = 4 * jdn + 139361631;
  j += Math.floor((Math.floor((4 * jdn + 183187720) / 146097) * 3) / 4) * 4 - 3908;
  const i = Math.floor((j % 1461) / 4) * 5 + 308;
  const gd = Math.floor((i % 153) / 5) + 1;
  const gm = (Math.floor(i / 153) % 12) + 1;
  const gy = Math.floor(j / 1461) - 100100 + Math.floor((8 - gm) / 6);
  return { gy, gm, gd };
}

function j2d(jy: number, jm: number, jd: number): number {
  const r = jalCal(jy);
  return g2d(r.gy, 3, r.march) + (jm - 1) * 31 - Math.floor(jm / 7) * (jm - 7) + jd - 1;
}

function d2j(jdn: number): ShamsiDate {
  const gy = d2g(jdn).gy;
  let jy = gy - 621;
  const r = jalCal(jy);
  const jdn1f = g2d(gy, 3, r.march);
  let k = jdn - jdn1f;

  if (k >= 0) {
    if (k <= 185) {
      return { year: jy, month: 1 + Math.floor(k / 31), day: (k % 31) + 1 };
    }
    k -= 186;
  } else {
    jy -= 1;
    k += 179;
    if (r.leap === 1) k += 1;
  }
  return { year: jy, month: 7 + Math.floor(k / 30), day: (k % 30) + 1 };
}

/** Parses an ISO date (YYYY-MM-DD) into its Solar Hijri equivalent. */
export function toShamsi(isoDate: string): ShamsiDate {
  const [y, m, d] = isoDate.split("-").map((v) => Number.parseInt(v, 10));
  return d2j(g2d(y, m, d));
}

export function toShamsiFromDate(date: Date): ShamsiDate {
  return d2j(g2d(date.getFullYear(), date.getMonth() + 1, date.getDate()));
}

/** Solar Hijri date -> ISO date string (YYYY-MM-DD). */
export function shamsiToIso(date: ShamsiDate): string {
  const g = d2g(j2d(date.year, date.month, date.day));
  return `${g.gy.toString().padStart(4, "0")}-${g.gm
    .toString()
    .padStart(2, "0")}-${g.gd.toString().padStart(2, "0")}`;
}

export function isShamsiLeapYear(year: number): boolean {
  return jalCal(year).leap === 0;
}

export function shamsiMonthLength(year: number, month: number): number {
  if (month <= 6) return 31;
  if (month <= 11) return 30;
  return isShamsiLeapYear(year) ? 30 : 29;
}

export function shamsiToday(): ShamsiDate {
  return toShamsiFromDate(new Date());
}

File: core/common/src/main/kotlin/app/worktrack/core/common/time/SolarHijri.kt
package app.worktrack.core.common.time

import java.time.LocalDate

/**
 * Solar Hijri (هجری شمسی) date — the official calendar of Afghanistan.
 * Month 1 is Hamal/حمل (vernal equinox, ~21 March).
 */
data class SolarHijriDate(val year: Int, val month: Int, val day: Int) {
    init {
        require(month in 1..12) { "month must be 1..12" }
        require(day in 1..31) { "day must be 1..31" }
    }

    fun toGregorian(): LocalDate = SolarHijri.toGregorian(this)

    /** Sortable "1405-04" style key for one month; used for paging state. */
    fun monthKey(): String = "%04d-%02d".format(year, month)
}

/**
 * Solar Hijri <-> Gregorian conversion using the arithmetic astronomical-cycle
 * algorithm from jalaali-js (Behrooz/Birashk break years), accurate for the
 * years this platform will ever process (1178–1633 AP / 1799–2254 AD).
 * Afghanistan shares the leap-year structure; only month names differ.
 */
object SolarHijri {

    private val BREAKS = intArrayOf(
        -61, 9, 38, 199, 426, 686, 756, 818, 1111, 1181, 1210,
        1635, 2060, 2097, 2192, 2262, 2324, 2394, 2456, 3178,
    )

    fun fromGregorian(date: LocalDate): SolarHijriDate = d2j(g2d(date.year, date.monthValue, date.dayOfMonth))

    fun toGregorian(date: SolarHijriDate): LocalDate = d2g(j2d(date.year, date.month, date.day))

    fun today(timeProvider: TimeProvider): SolarHijriDate = fromGregorian(timeProvider.today())

    fun isLeapYear(year: Int): Boolean = jalCal(year).leap == 0

    fun monthLength(year: Int, month: Int): Int = when {
        month <= 6 -> 31
        month <= 11 -> 30
        else -> if (isLeapYear(year)) 30 else 29
    }

    /** First Gregorian day of a Solar Hijri month (for date-range queries). */
    fun monthStart(year: Int, month: Int): LocalDate = toGregorian(SolarHijriDate(year, month, 1))

    fun monthEnd(year: Int, month: Int): LocalDate =
        toGregorian(SolarHijriDate(year, month, monthLength(year, month)))

    // ------------------------------------------------------------ internals

    private data class JalCal(val leap: Int, val gy: Int, val march: Int)

    private fun jalCal(jy: Int): JalCal {
        require(jy in (BREAKS.first() + 1) until BREAKS.last()) { "year $jy out of supported range" }
        val gy = jy + 621
        var leapJ = -14
        var jp = BREAKS[0]

        var jump = 0
        for (i in 1 until BREAKS.size) {
            val jm = BREAKS[i]
            jump = jm - jp
            if (jy < jm) break
            leapJ += jump / 33 * 8 + jump % 33 / 4
            jp = jm
        }
        var n = jy - jp

        leapJ += n / 33 * 8 + (n % 33 + 3) / 4
        if (jump % 33 == 4 && jump - n == 4) leapJ += 1

        val leapG = gy / 4 - (gy / 100 + 1) * 3 / 4 - 150
        val march = 20 + leapJ - leapG

        if (jump - n < 6) n = n - jump + (jump + 4) / 33 * 33
        var leap = ((n + 1) % 33 - 1) % 4
        if (leap == -1) leap = 4

        return JalCal(leap = leap, gy = gy, march = march)
    }

    private fun g2d(gy: Int, gm: Int, gd: Int): Int {
        var d = (gy + (gm - 8) / 6 + 100100) * 1461 / 4 +
            (153 * ((gm + 9) % 12) + 2) / 5 + gd - 34840408
        d = d - (gy + 100100 + (gm - 8) / 6) / 100 * 3 / 4 + 752
        return d
    }

    private fun d2g(jdn: Int): LocalDate {
        var j = 4 * jdn + 139361631
        j += (4 * jdn + 183187720) / 146097 * 3 / 4 * 4 - 3908
        val i = j % 1461 / 4 * 5 + 308
        val gd = i % 153 / 5 + 1
        val gm = i / 153 % 12 + 1
        val gy = j / 1461 - 100100 + (8 - gm) / 6
        return LocalDate.of(gy, gm, gd)
    }

    private fun j2d(jy: Int, jm: Int, jd: Int): Int {
        val r = jalCal(jy)
        return g2d(r.gy, 3, r.march) + (jm - 1) * 31 - jm / 7 * (jm - 7) + jd - 1
    }

    private fun d2j(jdn: Int): SolarHijriDate {
        val gy = d2g(jdn).year
        var jy = gy - 621
        val r = jalCal(jy)
        val jdn1f = g2d(gy, 3, r.march)
        var k = jdn - jdn1f

        if (k >= 0) {
            if (k <= 185) {
                return SolarHijriDate(jy, 1 + k / 31, k % 31 + 1)
            }
            k -= 186
        } else {
            jy -= 1
            k += 179
            if (r.leap == 1) k += 1
        }
        return SolarHijriDate(jy, 7 + k / 30, k % 30 + 1)
    }
}
